### PR TITLE
feat(checkbox): add checkbox group and validation states

### LINKED ISSUE #12645: Field validation errors: CheckboxGroup implementation
This issue is to track the dev implementation of field validation errors on a checkbox group.

Open questions:
- CheckboxGroup does not exist. Should it?
- The wrapper can be invalid, and the individual component can be invalid as well?
- Might need to update the checkbox stories to not use a manual fieldset/legend - instead use FormGroup
- Might have to update FormGroup to inspect children checkboxes based on their `invalid` or not

**States to be added**:
- Invalid (error)
- Warn
- Helper text

## Spec

Checkbox - White theme

### PR DESCRIPTION
**Pull request by aledavila**
Closes #12645 

Implementation for CheckboxGroup and validation states 

#### Changelog

**New**

- added new `CheckboxGroup` component
- added `helperText` `invalid` `readOnly` and `warn` to `CheckboxGroup` 
- - added `helperText` `invalid` and `warn` to `Checkbox` 

**Changed**

- Updated stories to apply new `CheckboxGroup`

#### Testing / Reviewing

Make sure props work as intended.
Make sure Checkbox validation styles don't show up when in group.

### CODE REVIEW

**Review by aagonzales — CHANGES_REQUESTED** [review]
Looks good, except the state icons are off my 1px. Should be 2px left and right (its 3px left, 1px right atm)

image

**Review by francinelucca — APPROVED** [review]
LGTM, do we need to add a test file for `CheckboxGroup` ? or would that be tracked elsewhere?

**Comment by aledavila** [comment]
Yup pushed now @francinelucca thanks for the reminder

File: packages/react/src/components/Checkbox/__tests__/Checkbox-test.js [hunk 1]
```diff
@@ -60,6 +60,95 @@ describe('Checkbox', () => {
...
+        labelText="Checkbox label"
+        id="checkbox-label-1"
+        helperText="Helper text"
+      />
+    );
+
+    expect(screen.getByText('Helper text')).toBeInTheDocument();
+  });
+
+  it('should be set data-invalid when invalid prop is true', () => {
```
Comment (francinelucca) [inline packages/react/src/components/Checkbox/__tests__/Checkbox-test.js]:
```suggestion
  it('should set data-invalid when invalid prop is true', () => {
```

File: packages/react/src/components/Checkbox/__tests__/Checkbox-test.js [hunk 2]
```diff
@@ -60,6 +60,95 @@ describe('Checkbox', () => {
...
+        id="checkbox-label-1"
+        invalid
+        invalidText="Invalid text"
+      />
+    );
+
+    expect(screen.getByText('Invalid text')).toBeInTheDocument();
+  });
+
+  it('should be respect readOnly prop', () => {
```
Comment (francinelucca) [inline packages/react/src/components/Checkbox/__tests__/Checkbox-test.js]:
```suggestion
  it('should respect readOnly prop', () => {
```

File: packages/react/src/components/Checkbox/__tests__/Checkbox-test.js [hunk 3]
```diff
@@ -60,6 +60,95 @@ describe('Checkbox', () => {
...
+        labelText="Checkbox label"
+        id="checkbox-label-1"
+        readOnly
+      />
+    );
+
+    expect(container.firstChild).toHaveClass(`cds--checkbox-wrapper--readonly`);
+  });
+
+  it('should be respect warn prop', () => {
```
Comment (francinelucca) [inline packages/react/src/components/Checkbox/__tests__/Checkbox-test.js]:
```suggestion
  it('should respect warn prop', () => {
```

File: packages/react/src/components/Checkbox/__tests__/Checkbox-test.js [hunk 4]
```diff
@@ -60,6 +60,95 @@ describe('Checkbox', () => {
...
+
+    const warnIcon = container.querySelector(
+      `svg.cds--checkbox__invalid-icon--warning`
+    );
+
+    expect(container.firstChild).toHaveClass(`cds--checkbox-wrapper--warning`);
+    expect(warnIcon).toBeInTheDocument();
+  });
+
+  it('should display warnText if message is true', () => {
```
Comment (francinelucca) [inline packages/react/src/components/Checkbox/__tests__/Checkbox-test.js]:
```suggestion
  it('should display warnText if warn prop is true', () => {
```

File: packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        className="test"
+        legendText="Checkbox heading"
+        helperText="Helper text"
+      />
+    );
+
+    expect(screen.getByText('Helper text')).toBeInTheDocument();
+  });
+
+  it('should be set data-invalid when invalid prop is true', () => {
```
Comment (francinelucca) [inline packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js]:
```suggestion
  it('should set data-invalid when invalid prop is true', () => {
```

File: packages/react/src/components/Checkbox/__tests__/Checkbox-test.js [hunk 6]
```diff
@@ -60,6 +60,95 @@ describe('Checkbox', () => {
...
+      />
+    );
+
+    expect(screen.getByRole('checkbox')).toHaveAttribute(
+      'data-invalid',
+      'true'
+    );
+  });
+
+  it('should display invalidText if message is true', () => {
```
Comment (francinelucca) [inline packages/react/src/components/Checkbox/__tests__/Checkbox-test.js]:
```suggestion
  it('should display invalidText if invalid prop is true', () => {
```

File: packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+          id="checkbox-label-1"
+        />
+        <Checkbox labelText={`Checkbox label`} id="checkbox-label-2" />
+      </CheckboxGroup>
+    );
+
+    expect(container.firstChild).toHaveAttribute('data-invalid', 'true');
+  });
+
+  it('should display invalidText if message is true', () => {
```
Comment (francinelucca) [inline packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js]:
```suggestion
  it('should display invalidText if invalid prop is true', () => {
```

File: packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  it('should set the id for legend based on legendId', () => {
+    render(<CheckboxGroup legendId="legend-testid" legendText="legendtest" />);
+
+    expect(screen.getByText('legendtest')).toHaveAttribute(
+      'id',
+      'legend-testid'
+    );
+  });
+
+  it('should be respect readOnly prop', () => {
```
Comment (francinelucca) [inline packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js]:
```suggestion
  it('should respect readOnly prop', () => {
```

File: packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        <Checkbox labelText={`Checkbox label`} id="checkbox-label-2" />
+      </CheckboxGroup>
+    );
+
+    expect(container.firstChild).toHaveClass(
+      `${prefix}--checkbox-group--readonly`
+    );
+  });
+
+  it('should be respect warn prop', () => {
```
Comment (francinelucca) [inline packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js]:
```suggestion
  it('should respect warn prop', () => {
```

File: packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+      `svg.${prefix}--checkbox__invalid-icon--warning`
+    );
+
+    expect(container.firstChild).toHaveClass(
+      `${prefix}--checkbox-group--warning`
+    );
+    expect(warnIcon).toBeInTheDocument();
+  });
+
+  it('should display warnText if message is true', () => {
```
Comment (francinelucca) [inline packages/react/src/components/CheckboxGroup/CheckboxGroup-test.js]:
```suggestion
  it('should display warnText if warn prop is true', () => {
```

**Review by francinelucca — APPROVED** [review]
Minor wording suggestions

**Comment by aledavila** [comment]
@tay1orjones @tw15egan ready for review. 

@aagonzales I think we said that the 3px was fine

**Review by aagonzales — APPROVED** [review]
Yup we measured it and 3px is actually correct due to the checkbox construction.

**Review by tw15egan — APPROVED** [review]
Looks awesome! 👍🏻 ✅

**Pull request merged**